lib/browser/browser.ts

```diff
@@ -210,48 +208,55 @@ class OktaAuthBrowser extends OktaAuthBase implements OktaAuth, SignoutAPI {
     this.authStateManager = new AuthStateManager(this);
   }
 
+  // TODO: deprecate in 5.0
   signIn(opts) {
-    const loginWithCredential = (opts) => {
-      opts = clone(opts || {});
-      const _postToTransaction = (options?) => {
-        delete opts.sendFingerprint;
-        return postToTransaction(this, '/api/v1/authn', opts, options);
-      };
-      if (!opts.sendFingerprint) {
-        return _postToTransaction();
-      }
-      return this.fingerprint()
-      .then(function(fingerprint) {
-        return _postToTransaction({
-          headers: {
-            'X-Device-Fingerprint': fingerprint
-          }
-        });
-      });
-    };
+    deprecate('This method will be deprecated in v5.0, please use signInWithCredentials() instead.');
 
-    const loginWithRedirect = async (opts: SigninWithRedirectOptions = {}) => {
-      if(this._pending.handleLogin) { 
-        // Don't trigger second round
-        return;
-      }
-  
-      this._pending.handleLogin = true;
-      this.setFromUri(opts.fromUri);
-      try {
-        if (this.options.onAuthRequired) {
-          return await this.options.onAuthRequired(this);
-        }
-        return await this.loginRedirect(undefined, opts);
-      } finally {
-        this._pending.handleLogin = null;
-      }
+    opts = clone(opts || {});
+    const _postToTransaction = (options?) => {
+      delete opts.sendFingerprint;
+      return postToTransaction(this, '/api/v1/authn', opts, options);
     };
+    if (!opts.sendFingerprint) {
+      return _postToTransaction();
+    }
+    return this.fingerprint()
+    .then(function(fingerprint) {
+      return _postToTransaction({
+        headers: {
+          'X-Device-Fingerprint': fingerprint
+        }
+      });
+    });
+  }
 
-    if (isSignInWithCredentialsOptions(opts)) {
-      return loginWithCredential(opts);
-    } else {
-      return loginWithRedirect(opts);
+  /**
+   * Alias method of signIn()
+   */ 
+  signInWithCredentials(opts: SignInWithCredentialsOptions) {
+    return this.signIn(opts);
+  }
+
+  async signInWithRedirect({ fromUri, ...additionalParams }: SigninWithRedirectOptions = {}) {
+    if(this._pending.handleLogin) { 
+      // Don't trigger second round
+      return;
+    }
+
+    this._pending.handleLogin = true;
+    const { scopes, responseType } = this.options;
+    try {
+      // Trigger default signIn redirect flow
+      if (fromUri) {
+        browserStorage.getSessionStorage().setItem(REFERRER_PATH_STORAGE_KEY, fromUri);
```

Comment:
I think a problem with removing `setFromUri` and setting it directly here in `signInWithRedirect` is how this works with `onAuthRequired` and the protected route guards. Only the route knows what the URI is (or should be), but it doesn't know what auth flow you want to do. The saved URI is only needed if doing a redirect flow, so there is sense in having it here. It's just that `onAuthRequired` isn't going to have that information or be able to pass it. Typically `onAuthRequired` goes to a new route, like `/login`. When the auth flow is complete we do not want to redirect back to `/login` we want to go back to the page we were navigating to before `/login`

Reply:
I think logically we want something like:

1) "begin auth flow": called by a protected route. this should take as an optional parameter the URL/route to be set after the flow is complete
2) "on auth required": used to begin the auth flow. This may call one of the signIn methods or redirect to another route which implements a signin form
3) "finish auth flow": after the flow is complete (tokens have been set) it will cleanup storage and set the URL/route if it had been previously set

This logic should work whether you use redirect, popup, or custom signin form.

Reply:
> The saved URI is only needed if doing a redirect flow

This method by design is doing single purpose default signIn redirect. for `onAuthRequired`, as we agreed, it can be a callback/prop of the downstream SDKs, which can be a replacement of this default flow.

Reply:
My initial idea is that the `onAuthRequired` should not be called inside `signInWithRedirect`, but it should be handled in the route guard. Then it would make sense to put `setFromUri` and `getFromUri` back for the downstream SDKs to use. And we still keep `signInWithRedirect` only do the default redirect login.